en/performance/practical-search-performance-guide.md

```diff
@@ -0,0 +1,2278 @@
+</pre>
+
+<pre style="display:none" data-test="exec">
+$ sleep 60
+</pre>
+
+Once up, we can try our multi-tag search again - now using the `fast-search` attribute:
+
+<div class="pre-parent">
+  <button class="d-icon d-duplicate pre-copy-button" onclick="copyPreContent(this)"></button>
+<pre data-test="exec" data-test-assert-contains='"rock":'>
+$ vespa query 'yql=select track_id, tags from track where tags contains "rock" or tags contains "pop"' \
+  'hits=1' 
+</pre>
+</div>
+
+Now the `querytime` will be a few milliseconds since Vespa has built index structures to support
+`fast-search` in the attribute. The downside of enabling `fast-search` is 
+increased memory usage and slightly reduced indexing throughput. See also
+[when to use fast-search for attributes](feature-tuning.html#when-to-use-fast-search-for-attribute-fields).
+
+Finally, for those interested in using `match:text` for searching multi-valued field types 
+like [weightedset](../reference/schema-reference.html#type:weightedset), see
+[searching multi-value fields](../searching-multi-valued-fields.html).
+
+## Multi-valued query operators
+
+In this section we look at [multi-value query operators](../multivalue-query-operators.html) 
+and their query performance characteristics. 
+Many real-world use cases needs search using structured queries with many values. 
+
+Let us assume that we have a process which have built a user profile where we have identified tags which
+the user has shown previous interest for. 
+For example *hard rock*, *rock*, *metal* and *finnish metal*. How can we retrieve and rank 
+results using that information? One way is to search and rank using
+the dotproduct between the sparse user profile representation and the track tags representation.
+
+We can start with the [dotProduct()](../reference/query-language-reference.html#dotproduct) query operator.
+To gain control of [ranking](../ranking.html) we need to add a `rank-profile` to our schema:
+
+<pre data-test="file" data-path="app/schemas/track.sd">
+schema track {
+
+  document track {
+
+    field track_id type string {
+      indexing: summary | attribute
+      rank: filter
+      match: word
+    }
+    field title type string {
+      indexing: summary | index
+      index: enable-bm25
+    }
+    field artist type string {
+      indexing: summary | index
+    }
+    field tags type weightedset&lt;string&gt; {
+      indexing: summary | attribute
+      attribute: fast-search
+    }
+    field similar type tensor&lt;float&gt;(trackid{}) {
+      indexing: summary | attribute
+    }
+  }
+
+  fieldset default {
+    fields: title, artist
+  }
+
+  document-summary track_id {
+    summary track_id type string { 
+      source: track_id
+    }
+  }
+  rank-profile personalized {
+    first-phase {
+      expression: rawScore(tags)
+    }
+  }
+}
+</pre>
+
+The `dotProduct`and `wand` query operators produces a rank feature called `rawScore(name)`, 
+we use this as our first-phase rank expression. 
+
```

Comment:
See [rawScore(field)](../reference/rank-features.html#rawScore(field)) for details.